File: benchmarking-suite/moPro/mopro-example-app/flutter/mopro_flutter_plugin/ios/Classes/MoproFlutterPlugin.swift
import Flutter
import Foundation
import UIKit

class FlutterG1 {
  let x: String
  let y: String
  let z: String

  init(x: String, y: String, z: String) {
    self.x = x
    self.y = y
    self.z = z
  }
}

class FlutterG2 {
  let x: [String]
  let y: [String]
  let z: [String]

  init(x: [String], y: [String], z: [String]) {
    self.x = x
    self.y = y
    self.z = z
  }
}

class FlutterCircomProof {
  let a: FlutterG1
  let b: FlutterG2
  let c: FlutterG1
  let `protocol`: String
  let curve: String

  init(a: FlutterG1, b: FlutterG2, c: FlutterG1, `protocol`: String, curve: String) {
    self.a = a
    self.b = b
    self.c = c
    self.`protocol` = `protocol`
    self.curve = curve
  }
}

class FlutterCircomProofResult {
  let proof: FlutterCircomProof
  let inputs: [String]

  init(proof: FlutterCircomProof, inputs: [String]) {
    self.proof = proof
    self.inputs = inputs
  }
}

func convertCircomProof(res: CircomProofResult) -> [String: Any] {
  let g1a = FlutterG1(x: res.proof.a.x, y: res.proof.a.y, z: res.proof.a.z)
  let g2b = FlutterG2(x: res.proof.b.x, y: res.proof.b.y, z: res.proof.b.z)
  let g1c = FlutterG1(x: res.proof.c.x, y: res.proof.c.y, z: res.proof.c.z)
  let circomProof = FlutterCircomProof(
    a: g1a, b: g2b, c: g1c, `protocol`: res.proof.protocol, curve: res.proof.curve)
  let circomProofResult = FlutterCircomProofResult(proof: circomProof, inputs: res.inputs)
  let resultMap: [String: Any] = [
    "proof": [
      "a": [
        "x": circomProofResult.proof.a.x,
        "y": circomProofResult.proof.a.y,
        "z": circomProofResult.proof.a.z,
      ],
      "b": [
        "x": circomProofResult.proof.b.x,
        "y": circomProofResult.proof.b.y,
        "z": circomProofResult.proof.b.z,
      ],
      "c": [
        "x": circomProofResult.proof.c.x,
        "y": circomProofResult.proof.c.y,
        "z": circomProofResult.proof.c.z,
      ],
      "protocol": circomProofResult.proof.protocol,
      "curve": circomProofResult.proof.curve,
    ],
    "inputs": circomProofResult.inputs,
  ]
  return resultMap
}

func convertCircomProofResult(proof: [String: Any]) -> CircomProofResult {
  let proofMap = proof["proof"] as! [String: Any]
  let aMap = proofMap["a"] as! [String: String]
  let g1a = G1(x: aMap["x"] ?? "0", y: aMap["y"] ?? "0", z: aMap["z"] ?? "1")
  let bMap = proofMap["b"] as! [String: [String]]
  let g2b = G2(
    x: bMap["x"] ?? ["1", "0"], y: bMap["y"] ?? ["1", "0"], z: bMap["z"] ?? ["1", "0"])
  let cMap = proofMap["c"] as! [String: String]
  let g1c = G1(x: cMap["x"] ?? "0", y: cMap["y"] ?? "0", z: cMap["z"] ?? "1")
  let circomProof = CircomProof(
    a: g1a, b: g2b, c: g1c, `protocol`: proofMap["protocol"] as! String,
    curve: proofMap["curve"] as! String)
  let circomProofResult = CircomProofResult(
    proof: circomProof, inputs: proof["inputs"] as! [String])
  return circomProofResult
}

public class MoproFlutterPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "mopro_flutter", binaryMessenger: registrar.messenger())
    let instance = MoproFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "generateCircomProof":
      guard let args = call.arguments as? [String: Any],
        let zkeyPath = args["zkeyPath"] as? String,
        let inputs = args["inputs"] as? String,
        let proofLib = args["proofLib"] as? Int
      else {
        result(
          FlutterError(
            code: "ARGUMENT_ERROR", message: "Missing arguments \(call.arguments)", details: nil))
        return
      }

      DispatchQueue.global(qos: .userInitiated).async {
        do {
          var moproProofLib: CircomProofLib
          if proofLib == 0 {
            moproProofLib = CircomProofLib.arkworks
          } else {
            moproProofLib = CircomProofLib.rapidsnark
          }
          // Call the function from mopro.swift
          let proofResult = try generateCircomProof(
            zkeyPath: zkeyPath, circuitInputs: inputs, proofLib: moproProofLib)
          let resultMap = convertCircomProof(res: proofResult)

          // Return the proof and inputs as a map supported by the StandardMethodCodec
          DispatchQueue.main.async {
            result(resultMap)
          }
        } catch {
          DispatchQueue.main.async {
            result(
              FlutterError(
                code: "PROOF_GENERATION_ERROR", message: "Failed to generate proof",
                details: error.localizedDescription))
          }
        }
      }

    case "verifyCircomProof":
      guard let args = call.arguments as? [String: Any],
        let zkeyPath = args["zkeyPath"] as? String,
        let proof = args["proof"] as? [String: Any],
        let proofLib = args["proofLib"] as? Int
      else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments \(call.arguments)", details: nil))
        return
      }

      DispatchQueue.global(qos: .userInitiated).async {
        do {
          var moproProofLib: CircomProofLib
          if proofLib == 0 {
            moproProofLib = CircomProofLib.arkworks
          } else {
            moproProofLib = CircomProofLib.rapidsnark
          }
          let circomProofResult = convertCircomProofResult(proof: proof)
          // Call the function from mopro.swift
          let valid = try verifyCircomProof(
            zkeyPath: zkeyPath, proofResult: circomProofResult, proofLib: moproProofLib)

          // Return the proof and inputs as a map supported by the StandardMethodCodec
          DispatchQueue.main.async {
            result(valid)
          }
        } catch {
          DispatchQueue.main.async {
            result(
              FlutterError(
                code: "PROOF_VERIFICATION_ERROR", message: "Failed to verify proof",
                details: error.localizedDescription))
          }
        }
      }


    case "generateNoirProof":
      guard let args = call.arguments as? [String: Any],
        let circuitPath = args["circuitPath"] as? String,
        let inputs = args["inputs"] as? [String],
        let onChain = args["onChain"] as? Bool,
        let vk = args["vk"] as? FlutterStandardTypedData,
        let lowMemoryMode = args["lowMemoryMode"] as? Bool
      else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
        return
      }

      let srsPath = args["srsPath"] as? String

      do {
        let proofResult = try generateNoirProof(
          circuitPath: circuitPath, srsPath: srsPath, inputs: inputs, onChain: onChain, vk: vk.data, lowMemoryMode: lowMemoryMode)
        result(proofResult)
      } catch {
        result(
          FlutterError(
            code: "PROOF_GENERATION_ERROR", message: "Failed to generate proof",
            details: error.localizedDescription))
      }

    case "verifyNoirProof":
      guard let args = call.arguments as? [String: Any],
        let circuitPath = args["circuitPath"] as? String,
        let proof = args["proof"] as? FlutterStandardTypedData,
        let onChain = args["onChain"] as? Bool,
        let vk = args["vk"] as? FlutterStandardTypedData,
        let lowMemoryMode = args["lowMemoryMode"] as? Bool
      else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
        return
      }

      do {
        let valid = try verifyNoirProof(circuitPath: circuitPath, proof: proof.data, onChain: onChain, vk: vk.data, lowMemoryMode: lowMemoryMode)
        result(valid)
      } catch {
        result(
          FlutterError(
            code: "PROOF_VERIFICATION_ERROR", message: "Failed to verify proof",
            details: error.localizedDescription))
      }
    case "getNoirVerificationKey":
      guard let args = call.arguments as? [String: Any],
        let circuitPath = args["circuitPath"] as? String,
        let onChain = args["onChain"] as? Bool,
        let lowMemoryMode = args["lowMemoryMode"] as? Bool
      else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
        return
      }

      let srsPath = args["srsPath"] as? String

      do {
        let vkResult = try getNoirVerificationKey(circuitPath: circuitPath, srsPath: srsPath, onChain: onChain, lowMemoryMode: lowMemoryMode)
        result(vkResult)
      } catch {
        result(
          FlutterError(
            code: "VK_GENERATION_ERROR", message: "Failed to generate verification key",
            details: error.localizedDescription))
      }
    case "generateCairoProof":
      guard let args = call.arguments as? [String: Any],
        let programJson = args["programJson"] as? String,
        let inputsJson = args["inputsJson"] as? String
      else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
        return
      }

      do {
        let proofOutput = try cairoProve(programJson: programJson, inputsJson: inputsJson)
        // Convert to map for Flutter
        let resultMap: [String: Any] = [
            "proof": proofOutput.proof
        ]
        result(resultMap)
      } catch {
        result(
          FlutterError(
            code: "CAIRO_PROOF_ERROR", message: "Failed to generate Cairo proof",
            details: error.localizedDescription))
      }

    case "verifyCairoProof":
      guard let args = call.arguments as? [String: Any],
        let proof = args["proof"] as? FlutterStandardTypedData
      else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
        return
      }

      do {
        let verifyOutput = try cairoVerify(proofBytes: proof.data)
        let resultMap: [String: Any] = [
            "is_valid": verifyOutput.isValid
        ]
        result(resultMap)
      } catch {
        result(
          FlutterError(
            code: "CAIRO_VERIFY_ERROR", message: "Failed to verify Cairo proof",
            details: error.localizedDescription))
      }

    case "getIOSMemoryUsage":
        let memoryInfo = getMemoryUsage()
        result(memoryInfo)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  func getMemoryUsage() -> [String: Int64] {
      var taskInfo = mach_task_basic_info()
      var count = mach_msg_type_number_t(MemoryLayout<mach_task_basic_info>.size) / 4
      let kerr: kern_return_t = withUnsafeMutablePointer(to: &taskInfo) {
          $0.withMemoryRebound(to: integer_t.self, capacity: 1) {
              task_info(mach_task_self_, task_flavor_t(MACH_TASK_BASIC_INFO), $0, &count)
          }
      }

      var usedMemory: Int64 = 0
      if kerr == KERN_SUCCESS {
          usedMemory = Int64(taskInfo.resident_size)
      }

      let totalMemory = Int64(ProcessInfo.processInfo.physicalMemory)
      
      return ["used": usedMemory, "total": totalMemory]
  }
}

File: benchmarking-suite/moPro/mopro-example-app/flutter/mopro_flutter_plugin/android/src/main/kotlin/com/example/mopro_flutter/MoproFlutterPlugin.kt
package com.example.mopro_flutter

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

import uniffi.mopro.*

import io.flutter.plugin.common.StandardMethodCodec

class FlutterG1(x: String, y: String, z: String) {
    val x = x
    val y = y
    val z = z
}

class FlutterG2(x: List<String>, y: List<String>, z: List<String>) {
    val x = x
    val y = y
    val z = z
}

class FlutterGroth16Proof(a: FlutterG1, b: FlutterG2, c: FlutterG1, protocol: String, curve: String) {
    val a = a
    val b = b
    val c = c
    val protocol = protocol
    val curve = curve
}

class FlutterGroth16ProofResult(proof: FlutterGroth16Proof, inputs: List<String>) {
    val proof = proof
    val inputs = inputs
}

fun convertGroth16Proof(res: Groth16ProofResult): Map<String, Any> {
    val g1a = FlutterG1(res.proof.a.x, res.proof.a.y, res.proof.a.z)
            val g2b = FlutterG2(res.proof.b.x, res.proof.b.y, res.proof.b.z)
            val g1c = FlutterG1(res.proof.c.x, res.proof.c.y, res.proof.c.z)
            val circomProof = FlutterGroth16Proof(g1a, g2b, g1c, res.proof.protocol, res.proof.curve)
            val circomProofResult = FlutterGroth16ProofResult(circomProof, res.inputs)
            // Convert to Map before sending
    val resultMap = mapOf(
        "proof" to mapOf(
            "a" to mapOf(
                "x" to circomProofResult.proof.a.x,
                "y" to circomProofResult.proof.a.y,
                "z" to circomProofResult.proof.a.z
            ),
            "b" to mapOf(
                "x" to circomProofResult.proof.b.x,
                "y" to circomProofResult.proof.b.y,
                "z" to circomProofResult.proof.b.z
            ),
            "c" to mapOf(
                "x" to circomProofResult.proof.c.x,
                "y" to circomProofResult.proof.c.y,
                "z" to circomProofResult.proof.c.z
            ),
            "protocol" to circomProofResult.proof.protocol,
            "curve" to circomProofResult.proof.curve
        ),
        "inputs" to circomProofResult.inputs
    )
    return resultMap
}
fun convertGroth16ProofResult(proofResult: Map<String, Any>): Groth16ProofResult {
    val proofMap = proofResult["proof"] as Map<String, Any>
    val aMap = proofMap["a"] as Map<String, Any>
    val g1a = G1(
        aMap["x"] as String,
        aMap["y"] as String,
        aMap["z"] as String
    )
    val bMap = proofMap["b"] as Map<String, Any>
    val g2b = G2(
        bMap["x"] as List<String>,
        bMap["y"] as List<String>,
        bMap["z"] as List<String>
    )
    val cMap = proofMap["c"] as Map<String, Any>
    val g1c = G1(
        cMap["x"] as String,
        cMap["y"] as String,
        cMap["z"] as String
    )
    val groth16Proof = Groth16Proof(
        g1a,
        g2b,
        g1c,
        proofMap["protocol"] as String,
        proofMap["curve"] as String
    )
    val groth16ProofResult = Groth16ProofResult(groth16Proof, proofResult["inputs"] as List<String>)
    return groth16ProofResult
  }

/** MoproFlutterPlugin */
class MoproFlutterPlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(
            flutterPluginBinding.binaryMessenger,
            "mopro_flutter",
            StandardMethodCodec.INSTANCE
        )
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "generateGroth16Proof") {
            val zkeyPath = call.argument<String>("zkeyPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing zkeyPath",
                null
            )

            val inputs =
                call.argument<String>("inputs") ?: return result.error(
                    "ARGUMENT_ERROR",
                    "Missing inputs",
                    null
                )
            
            val proofLibIndex = call.argument<Int>("proofLib") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing proofLib",
                null
            )

            val proofLib = if (proofLibIndex == 0) ProofLib.ARKWORKS else ProofLib.RAPIDSNARK

            try {
                val res = generateGroth16Proof(zkeyPath, inputs, proofLib)
                val resultMap = convertGroth16Proof(res)
                result.success(resultMap)
            } catch (e: Exception) {
                result.error("PROOF_GENERATION_ERROR", "Failed to generate Groth16 proof", e.message)
            }
        } else if (call.method == "verifyGroth16Proof") {
            val zkeyPath = call.argument<String>("zkeyPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing zkeyPath",
                null
            )

            val proof = call.argument<Map<String, Any>>("proof") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing proof",
                null
            )

            val proofLibIndex = call.argument<Int>("proofLib") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing proofLib",
                null
            )

            val proofLib = if (proofLibIndex == 0) ProofLib.ARKWORKS else ProofLib.RAPIDSNARK

            try {
                val circomProofResult = convertGroth16ProofResult(proof)
                val res = verifyGroth16Proof(zkeyPath, circomProofResult, proofLib)
                result.success(res)
            } catch (e: Exception) {
                result.error("PROOF_VERIFICATION_ERROR", "Failed to verify Groth16 proof", e.message)
            }

        } else if (call.method== "generateBarretenbergProof") {
            val circuitPath = call.argument<String>("circuitPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing circuitPath",
                null
            )

            val srsPath = call.argument<String>("srsPath") 

            val inputs = call.argument<List<String>>("inputs") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing inputs",
                null
            )

            val onChain = call.argument<Boolean>("onChain") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing onChain",
                null
            )

            val vk = call.argument<ByteArray>("vk") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing vk",
                null
            )

            val lowMemoryMode = call.argument<Boolean>("lowMemoryMode") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing lowMemoryMode",
                null
            )

            val res = generateBarretenbergProof(circuitPath, srsPath, inputs, onChain, vk, lowMemoryMode)
            result.success(res)
        } else if (call.method== "verifyBarretenbergProof") {
            val circuitPath = call.argument<String>("circuitPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing circuitPath",
                null
            )

            val proof = call.argument<ByteArray>("proof") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing proof",
                null
            )

            val onChain = call.argument<Boolean>("onChain") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing onChain",
                null
            )

            val vk = call.argument<ByteArray>("vk") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing vk",
                null
            )

            val lowMemoryMode = call.argument<Boolean>("lowMemoryMode") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing lowMemoryMode",
                null
            )

            val res = verifyBarretenbergProof(circuitPath, proof, onChain, vk, lowMemoryMode)
            result.success(res)

        } else if (call.method== "getBarretenbergVerificationKey") {
            val circuitPath = call.argument<String>("circuitPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing circuitPath",
                null
            )

            val srsPath = call.argument<String>("srsPath")

            val onChain = call.argument<Boolean>("onChain") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing onChain",
                null
            )

            val lowMemoryMode = call.argument<Boolean>("lowMemoryMode") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing lowMemoryMode",
                null
            )

            val res = getBarretenbergVerificationKey(circuitPath, srsPath, onChain, lowMemoryMode)
            result.success(res)

        } else if (call.method== "generateRisc0Proof") {
            val input = call.argument<Int>("input") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing input",
                null
            )

            try {
                val res = risc0Prove(input.toUInt())
                val resultMap = mapOf(
                    "receipt" to res.receipt
                )
                result.success(resultMap)
            } catch (e: Exception) {
                result.error("PROOF_GENERATION_ERROR", "Failed to generate RISC0 proof", e.message)
            }

        } else if (call.method== "verifyRisc0Proof") {
            val receiptBytes = call.argument<ByteArray>("receiptBytes") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing receiptBytes",
                null
            )

            try {
                val res = risc0Verify(receiptBytes)
                val resultMap = mapOf(
                    "isValid" to res.isValid,
                    "outputValue" to res.outputValue.toInt()
                )
                result.success(resultMap)
            } catch (e: Exception) {
                result.error("PROOF_VERIFICATION_ERROR", "Failed to verify RISC0 proof", e.message)
            }
            
            
        } else if (call.method == "generateCairoProof") {
            val programJson = call.argument<String>("programJson") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing programJson",
                null
            )
            val inputsJson = call.argument<String>("inputsJson") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing inputsJson",
                null
            )

            try {
                val res = cairoProve(programJson, inputsJson)
                val resultMap = mapOf(
                    "proof" to res.proof
                )
                result.success(resultMap)
            } catch (e: Exception) {
                result.error("CAIRO_PROOF_ERROR", "Failed to generate Cairo proof: ${e.message}", null)
            }

        } else if (call.method == "verifyCairoProof") {
            val proof = call.argument<ByteArray>("proof") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing proof",
                null
            )

            try {
                val res = cairoVerify(proof)
                val resultMap = mapOf(
                    "is_valid" to res.isValid
                )
                result.success(resultMap)
            } catch (e: Exception) {
                result.error("CAIRO_VERIFY_ERROR", "Failed to verify Cairo proof: ${e.message}", null)
            }

        } else if (call.method == "generateProveKitProof") {
            val proverPath = call.argument<String>("proverPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing proverPath",
                null
            )
            val inputToml = call.argument<String>("inputToml") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing inputToml",
                null
            )

            try {
                val res = provekitProve(proverPath, inputToml)
                val resultMap = mapOf(
                    "proof" to res.proof
                )
                result.success(resultMap)
            } catch (e: Exception) {
                result.error("PROVEKIT_PROOF_ERROR", "Failed to generate ProveKit proof: ${e.message}", null)
            }

        } else if (call.method == "verifyProveKitProof") {
            val verifierPath = call.argument<String>("verifierPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing verifierPath",
                null
            )
            val proof = call.argument<ByteArray>("proof") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing proof",
                null
            )

            try {
                val res = provekitVerify(verifierPath, proof)
                val resultMap = mapOf(
                    "is_valid" to res.isValid
                )
                result.success(resultMap)
            } catch (e: Exception) {
                result.error("PROVEKIT_VERIFY_ERROR", "Failed to verify ProveKit proof: ${e.message}", null)
            }

        } else {
            result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}
